src/overlays/actors/ovl_Bg_Ddan_Jd/z_bg_ddan_jd.c

```diff
@@ -31,17 +33,157 @@ const ActorInit Bg_Ddan_Jd_InitVars = {
+            this->targetHeight = this->dyna.actor.initPosRot.pos.y + MOVE_HEIGHT_MIDDLE;
+        }
+
+        this->actionFunc = BgDdanJd_Move;
+    }
+}
+
+// Handles dust particles and sfx when moving
+void BgDdanJd_MoveEffects(BgDdanJd* this, GlobalContext* globalCtx) {
+    Vec3f dustPos;
+
+    // Generate random dust particles at the platform's base.
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ddan_Jd/func_80870B88.s")
+    dustPos.y = this->dyna.actor.initPosRot.pos.y;
+    if ((globalCtx->gameplayFrames & 1) != 0) {
+        dustPos.x = this->dyna.actor.posRot.pos.x + 65.0f;
+        dustPos.z = Math_Rand_CenteredFloat(110.0f) + this->dyna.actor.posRot.pos.z;
+        func_80033480(globalCtx, &dustPos, 5.0f, 1, 0x14, 0x3C, 1);
+        dustPos.x = this->dyna.actor.posRot.pos.x - 65.0f;
+        dustPos.z = Math_Rand_CenteredFloat(110.0f) + this->dyna.actor.posRot.pos.z;
+        func_80033480(globalCtx, &dustPos, 5.0f, 1, 0x14, 0x3C, 1);
+    } else {
+        dustPos.x = Math_Rand_CenteredFloat(110.0f) + this->dyna.actor.posRot.pos.x;
+        dustPos.z = this->dyna.actor.posRot.pos.z + 65.0f;
+        func_80033480(globalCtx, &dustPos, 5.0f, 1, 0x14, 0x3C, 1);
+        dustPos.x = Math_Rand_CenteredFloat(110.0f) + this->dyna.actor.posRot.pos.x;
+        dustPos.z = this->dyna.actor.posRot.pos.z - 65.0f;
+        func_80033480(globalCtx, &dustPos, 5.0f, 1, 0x14, 0x3C, 1);
+    }
+    if (this->stepSpeed == SHORTCUT_STEP_SPEED) {
+        func_8002F974(&this->dyna.actor, NA_SE_EV_ELEVATOR_MOVE - SFX_FLAG);
+    }
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ddan_Jd/func_80870D2C.s")
+// Implements the platform's movement state
+void BgDdanJd_Move(BgDdanJd* this, GlobalContext* globalCtx) {
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ddan_Jd/func_80870F00.s")
+    // if this is the platform that rises all the way to the top, and the switch state has changed to 1
+    if (this->stepSpeed == 1 && this->dyna.actor.params < 0x40 &&
+        Flags_GetSwitch(globalCtx, this->dyna.actor.params) != 0) {
+        this->stepSpeed = SHORTCUT_STEP_SPEED;
+        this->state = STATE_GO_MIDDLE_FROM_BOTTOM;
+        this->dyna.actor.posRot.pos.y = this->dyna.actor.initPosRot.pos.y + MOVE_HEIGHT_MIDDLE;
+        this->idleTimer = 0;
+        this->actionFunc = BgDdanJd_Idle;
+        func_800800F8(globalCtx, 0xBF4, -0x63, &this->dyna.actor, 0);
+    } else {
+        if (Math_ApproxF(&this->dyna.actor.posRot.pos.y, this->targetHeight, (f32)this->stepSpeed)) {
+            Audio_PlayActorSound2(&this->dyna.actor, NA_SE_EV_PILLAR_MOVE_STOP);
+            this->actionFunc = BgDdanJd_Idle;
+        }
+    }
+    BgDdanJd_MoveEffects(this, globalCtx);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Ddan_Jd/BgDdanJd_Update.s")
+void BgDdanJd_Update(Actor* thisx, GlobalContext* globalCtx) {
+    BgDdanJd* this = THIS;
+    this->actionFunc(thisx, globalCtx);
```

Comment:
pass `this` instead of `thisx` if possible?

Reply:
in this case it should definitely be this, cause action funcs take the casted actor